### PR TITLE
Add destroy command

### PR DESCRIPTION
**Pull request by antonmoiseev**
I just started working on this (#80). It currently takes more exploration than actual code writing, but I'd rather open the PR early on to have a chance to ask questions, and hopefully get the feedback while it's still very little code to review.

Several questions that I have so far:

1. Probably the biggest concern currently is reusing `files()` from the corresponding `generate` commands. `files()` reads actual files content from disk and renders the templates, which is completely unnecessary for `destroy` command. Should we consider refactoring this and splitting file paths resolving and template rendering into separate functions? It could be done as another issue after the initial version of `destroy` is done.

2. <details>
      <summary>Do you have a preference which word to use for console messages: "destroy", "delete" or "remove" (expand to see examples)?</summary>

      - Destroying component files...
      - Removing component files...
      - Deleting component files...
    </details>

3. After deleting generated files the directory might or might not be empty (users can create files manually). Do you think we should get rid of empty directories as well?

4. I find it ambiguous to refer to different types of entities that CLI can generate/destroy as "components":

    image

    Have you discussed this previously? Did you decide to settle on this term for now?

### CODE REVIEW

**Comment by antonmoiseev** [comment]
Just ran into another issue with reusing `files()` from `generate` commands. Since `files()` renders the templates `templateVars` that often come from `generate` builder flags must be defined. To avoid defining fake builder flags for `destroy` commands I came up with a workaround - wrapping `files()` into another function, so we can pass missing `templateVars` that the templates expect. See an example [here](https://github.com/redwoodjs/redwood/pull/487/commits/4bc1bdc319e2a8e4caa5f47cd71e0615826da606#diff-c2aeaa056ef4a69b0e0ed7b964ecf7ebR4-R14).

**Comment by jonniebigodes** [comment]
@antonmoiseev wow this is a good one. If you allow me my two cents here. I think it would be best to use "remove" instead of destroy. Wording wise it would be a better approach

**Comment by cannikin** [comment]
@antonmoiseev Thanks so much for doing this PR!

1. Agree, we should refactor the generators to have a function that just returns the list of files and not their contents. This will help with the #82 feature as well. But yeah that can probably wait until after this is merged and working.

2. I've been modeling the generators from those in Rails and they use "destroy" so that's what my brain defaulted to, but I'm open to another synonym if people like another one better. Can I throw "ungenerate" into the ring? 😀

3. I think we can delete an empty directory, yeah.

4. Interesting, yeah "component component" isn't right, it should just be "component". We do want to append it to the other things though, like "layout component", "page component", etc. The quick way would be to override the `desc` property that comes out of the `createYargsForComponentGeneration` call in the component generator.

**Comment by mojombo** [comment]
I think "destroy" is a good name, as it's quite dangerous sounding, which reflects the truth of the command. You should think twice before you call it.

**Comment by antonmoiseev** [comment]
@cannikin thank you for the feedback! Sorry for "ungenerate" though 😬, left "destroy" for now as Tom suggested.

Just to give you an update on the progress - all commands seem to work fine now, but I need to get rid of empty dirs and add tests. One thing I'm not sure about is `scaffold.css` file. When destroying a scaffold there is no easy way to identify whether anyone else is still using `scaffold.css` file, so I cannot decide whether it needs to be cleaned up as well. I was thinking about searching for `rw-scaffold` CSS class within layout components, but probably it would be too much. Do you have any suggestions here?

**Comment by cannikin** [comment]
I think it’s okay to leave it. Checking if anything else uses it seems like more trouble than it’s worth.

**Comment by antonmoiseev** [comment]
Well, I now clean up empty dirs after `destroy` runs but feel free to let me know if you find it too verbose. Here is probably an extreme version of `destroy` log:

image

**Comment by cannikin** [comment]
That log looks fine to me, I like seeing everything it's doing!

**Comment by mojombo** [comment]
Agreed, you definitely want a record of everything being deleted.

**Comment by thedavidprice** [comment]
Hi @antonmoiseev is this one ready for review + merge? Please disable "draft" and loop in Rob Cameron if so. Thanks!

**Comment by antonmoiseev** [comment]
Feature-wise it's ready for review, but tests are missing. Destroy commands primarily consist of two parts:
1. Figuring out which files need to be deleted / reverted.
2. Deleting and reverting files.

Since destroy reuses `files()` from corresponding generate commands the first part is already covered by `generate` tests. To test the second part I need to run tests against actual file system, but it looks like you do not currently do this. So, the only idea that I came up with is to test [`removeRoutesFromRouterTask()`](https://github.com/redwoodjs/redwood/pull/487/commits/1e4df029866b78c8dc09f2e7befee0664d9a2dd5#diff-e07755b44b941719cef54bc305d02667R212-R228).

@cannikin could you please review this PR? If you have any tests suggestions, please let me know, I'll address them.

**Comment by cannikin** [comment]
Hmmm... @peterp can you think of any way to test that files are being deleted? We discussed mocking out the filesystem at one point, but when we went to just testing the result of `files()` rather than the generator itself (which would actually write files) I think we abandoned that approach.

In Ruby there's a gem called **mocha** that lets you set expectations on method calls. So you can say "`unlinkSync` should be called one time, and with the following argument(s)". It will then "absorb" that method call and not really do it, so you don't get any errors if the files don't actually exist. Is there similar functionality in Jest? The nomenclature for testing is all over the map, but in mocha these are called "mocks". There are also "stubs" which allow you to redefine what a method returns.

**Comment by thedavidprice** [comment]
@RobertBroersma any test-related suggestions here?

**Comment by RobertBroersma** [comment]
@thedavidprice I'm not familiar with the `cli` package so much, but in general I like to keep my tests as close to the user as possible. So for commands I would for every use-case:
1. Arrange Code/Files/Whatever the command deals with.
2. Execute the command (possibly through `exec` but probably just call the function behind yarg)
3. Assert the Code/Files/Whatever the command deals with is now correct.

@cannikin  There certainly is something like that in Jest! There's spies (for "spying" on functions, but leaving them intact), and mocks, for faking out functions.

The Jest docs on mocks actually feature an example for mocking `fs`, which would be the only thing I would indeed mock: https://jestjs.io/docs/en/manual-mocks

Hope that helps!

**Comment by antonmoiseev** [comment]
There is [mock-fs](https://github.com/tschaub/mock-fs) package that can mock entire `fs` module. But I believe you mentioned in the community forum that you prefer to mock as little as possible. In Dart ecosystem there is [test_descriptor](https://github.com/dart-lang/test_descriptor) package that allows defining and verifying directory structures. It crates directories in the system's temp folder and teardowns everything after the test run. I was hoping to find something similar in node.js as well.

**Comment by RobertBroersma** [comment]
I don't have experience with testing stuff on the file system, but if it's workable without mocking it that would be cool! I would be surprised if there are no packages available for that!

**Comment by cannikin** [comment]
I have no idea what to do about testing these methods that are all wrapped up in `Listr` calls, or the actual generators themselves. For now we took the easy path and just tested that `files()` returned from each generator contains the proper list of file paths and contents—we sort of cross our fingers and hope that everything else around the generators does their jobs. 😬Unfortunately none of us are JS testing experts and know of a good way to

As @RobertBroersma said:

> There's spies (for "spying" on functions, but leaving them intact), and mocks, for faking out functions.

If I was writing this in Ruby I would do exactly this—mock out the file system calls (like `fs.unlinkSync()`) and verify that the call was made with the proper argument (the path of the file to be deleted). No need to create fake files and have the OS actually delete them, we can be reasonably sure that will work as expected. We just want to make sure *our* code is making the proper calls. Maybe you can mock out that file system call and then invoke the destroyer's `handler()` function, and everything will magically work?

**Comment by RobertBroersma** [comment]
@cannikin @antonmoiseev  Another Jest feature that could be useful for template testing is snapshots! https://jestjs.io/docs/en/snapshot-testing

Instead of writing the file to the file system you keep a snapshot of the correct output. 

This doesn't really make sense for deleting files I guess. For that I agree it's probably easier to mock out the filesystem.

**Comment by thedavidprice** [comment]
>I have no idea what to do about testing these methods that are all wrapped up in Listr calls, or the actual generators themselves.

Agreed this is messy. And I'm super guilty of this re: `yarn-redwood-create.js` 😬The Listr implementation is a leading reason there's a lack of correlating tests. And I've been thinking it might be best practice for us to create functions outside Listr, which would then be called as needed per task. (Am I correct that something like this would move us a step closer to testing @cannikin?)

But this conversation made me curious about how/if you can isolate and run specific tasks within the [Listr array](https://github.com/SamVerschueren/listr#listrtasks-options). And, indeed, you can. The following runs the first task without any of the Listr elements (e.g. no display title, etc):
```javascript
const tasks = new Listr([ ... ])

tasks._tasks[0].run()
```

**Does isolating each task like this help with testing?**

_FYI_ --> Here’s_ the Listr structure for an array of three tasks with titles "First", "Second", "Third":
```javascript
[
  Task {
    _isScalar: false,
    observers: [],
    closed: false,
    isStopped: false,
    hasError: false,
    thrownError: null,
    _listr: Listr {
      _options: [Object],
      _tasks: [Circular],
      concurrency: 1,
      _RendererClass: [Function: UpdateRenderer],
      exitOnError: undefined
    },
    _options: {
      showSubtasks: true,
      concurrent: false,
      renderer: 'default',
      nonTTYRenderer: 'verbose'
    },
    _subtasks: [],
    _enabledFn: undefined,
    _isEnabled: true,
    output: undefined,
    title: 'First',
    skip: [Function: defaultSkipFn],
    task: [Function: task]
  },
  Task {
    _isScalar: false,
    observers: [],
    closed: false,
    isStopped: false,
    hasError: false,
    thrownError: null,
    _listr: Listr {
      _options: [Object],
      _tasks: [Circular],
      concurrency: 1,
      _RendererClass: [Function: UpdateRenderer],
      exitOnError: undefined
    },
    _options: {
      showSubtasks: true,
      concurrent: false,
      renderer: 'default',
      nonTTYRenderer: 'verbose'
    },
    _subtasks: [],
    _enabledFn: undefined,
    _isEnabled: true,
    output: undefined,
    title: 'Second',
    skip: [Function: defaultSkipFn],
    task: [Function: task]
  },
  Task {
    _isScalar: false,
    observers: [],
    closed: false,
    isStopped: false,
    hasError: false,
    thrownError: null,
    _listr: Listr {
      _options: [Object],
      _tasks: [Circular],
      concurrency: 1,
      _RendererClass: [Function: UpdateRenderer],
      exitOnError: undefined
    },
    _options: {
      showSubtasks: true,
      concurrent: false,
      renderer: 'default',
      nonTTYRenderer: 'verbose'
    },
    _subtasks: [],
    _enabledFn: undefined,
    _isEnabled: true,
    output: undefined,
    title: 'Third',
    skip: [Function: defaultSkipFn],
    task: [Function: task]
  }
]
```

**Comment by cannikin** [comment]
It seems like being able to test the individual Listr commands would help...for the destroy one you'd still need the files to exist for them to actually successfully delete, though. I feel like just mocking out `fs.unlinkSync` is the simplest thing we can do in this case. For the task that edits the Routes file we'd need to have a real one somewhere and then compare the before and after to see if they were edited correctly.

But yeah testing the Listr tasks seem like it would help for all the other stuff that aren't covered by any tests currently. We could go up a mocking level and just test that the Listr tasks call the correct methods in the generators, rather than testing the whole generator stack again. But too much of that and it starts getting pretty brittle...change anything anywhere and you've got this huge cascade of mocks to fix. Ugh.

@antonmoiseev how do you feel about mocking out the file system calls as described [here](https://jestjs.io/docs/en/manual-mocks#mocking-node-modules) and just checking that `unlinkSync` is called with the proper paths? I'm not sure how actually invoke the generator though...can you just invoke `handler()` from, for example, `commands/destroy/page/page.js`? I wonder if the fact that it's `async` causes any additional issues...

**Comment by antonmoiseev** [comment]
> how do you feel about mocking out the file system calls as described here and just checking that `unlinkSync` is called with the proper paths?

@cannikin Sounds good, that's what I tried, but was mocking `readFile()` and `writeFile()` from `lib/index.js`. Not sure though why you would want to invoke generators? Since we mock file system calls we don't need files to exist, we can just pass the name and verify that destroy attempts to delete the right files. I'll add the tests over this weekend.

**Comment by antonmoiseev** [comment]
I've added a [test for destroy page](https://github.com/redwoodjs/redwood/pull/487/commits/6a2bf2f576efbd7fa4b026034fdef16754336a25) command. It looks a bit differently from what I expected after our discussion above. @cannikin could you please take a look and let me know whether you want to proceed with this approach?

I tried to find the balance between mocking as little as possible and staying away from being too fragile:
* Probably the most controversial part is running the entire `Listr` task in a single test. I like it treats `Listr` as a black box instead of breaking into private APIs like `_tasks` lists. However we lose granularity - verifying deleted files and edited routes in a single test.
* I had to mock more `fs` functions than I expected for a couple of reasons:
    * To reliably test destroy command I had to compare files that _current_ `generate` implementation produces with files `destroy` attempts to delete. So, I had to mock APIs required to run `files` from corresponding `generate` command.
    * The only way to get results from most `Listr` tasks is to let them write into filesystem. Then we can read and verify results.

File: packages/cli/__mocks__/fs.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import { isString } from 'lodash'
+
+const fs = {
+  ...require.requireActual('fs'),
+}
```
Comment (antonmoiseev) [inline packages/cli/__mocks__/fs.js:5]: I tried this instead:
```js
const fs = jest.genMockFromModule('fs')
```
But it breaks in a weird way:
image

File: packages/cli/src/commands/destroy/page/__tests__/page.test.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+      '<Routes>',
+      '  <Route path="/about" page={AboutPage} name="about" />',
+      '  <Route path="/" page={HomePage} name="home" />',
+      '  <Route notfound page={NotFoundPage} />',
+      '</Routes>',
+    ].join('\n'),
+  })
+})
+
+test('destroys a page and route', async () => {
```
Comment (cannikin) [inline packages/cli/src/commands/destroy/page/__tests__/page.test.js:31]: How gross would it be to split this into two tests? Can you do separate "destroy page" and a "destroy routes" tests? And just mock the page stuff in the page test and the routes stuff in the route test?

Comment (cannikin) [reply packages/cli/src/commands/destroy/page/__tests__/page.test.js:31]: Will the destroyers just keep going if something they intend to destroy doesn't exist any more? Specifically, if you only mock the page, will the route destroyer fail, or vice versa?

File: packages/cli/src/commands/destroy/page/__tests__/page.test.js [hunk 3]
```diff
@@ -28,18 +28,40 @@ beforeEach(() => {
...
 
-test('destroys a page and route', async () => {
+test('destroys page files', async () => {
   const unlinkSpy = jest.spyOn(fs, 'unlinkSync')
   const t = tasks({ name: 'About' })
   t.setRenderer('silent')
 
-  return t.run().then(() => {
-    // Make sure all generated files were destroyed.
+  return t._tasks[0].run().then(() => {
```
Comment (antonmoiseev) [inline packages/cli/src/commands/destroy/page/__tests__/page.test.js:36]: @cannikin actually pretty straightforward! Didn't realize individual sub-tasks are also wrapped into `Task` class and runnable.

Destroyers never fail, they just skip the work that cannot be done. We can safely call them multiple times in a row for the same component.

**Comment by cannikin** [comment]
@antonmoiseev Breaking up that test into individuals looks great to me, if you want to add tests for the other generators then we should be good to go!

**Comment by cannikin** [comment]
@antonmoiseev Don't be mad but we just merged this: https://github.com/redwoodjs/redwood/pull/423 😬

**Comment by jtoar** [comment]
@antonmoiseev I was the author of #423 and have been trying to keep up with this PR as well--happy to help if you have any questions.

**Comment by antonmoiseev** [comment]
@cannikin @jtoar no worries! I had to merge master into this one to add destroyer for the new function generator anyway. I'll get back to you if I have questions on #423, but planning to finish the tests by tomorrow.

**Comment by antonmoiseev** [comment]
@jtoar @cannikin I just [adjusted](https://github.com/redwoodjs/redwood/pull/487/commits/f4ac09c681cda7c5a98dc6a8775c2a1e426a21d2) `destroy` to #423 changes and wondering whether it should support `--path` as well?

I'm genuinely happy we can generate namespaced scaffolds now, but frankly, I find it excessive and a bit confusing to support two ways to generate nested paths. I think it unnecessary expands API surface, makes users take inessential decisions, and slightly complicates the code. I may totally miss an important reason why it's there, if so, please let me know, otherwise I would suggest to drop it.

**Comment by jtoar** [comment]
Going to loop @thedavidprice and @olance in here. I should've prompted more of a discussion around this when it came up in #423. We didn't have an explicit important reason to include both, other than that including both was easy implementation-wise. I'm open to removing it. People seemed to like the path/model syntax more. Thoughts?

**Comment by thedavidprice** [comment]
### Clarification about Destroy supporting --path
@antonmoiseev Thanks again for all the amazing work on this. Question of clarification about:
> I just adjusted destroy to #423 changes and wondering _whether it should support --path as well?_

^^ is the "it" you're referring to the Destroy command? If so, how would supporting --path differ from your current implementation.

### Scaffold --Path option(s) discussion
@cannikin Definitely want to get your opinion on the question about `yarn rw g scaffold ...` being able to take either of the following as options:
1. `yarn rw g scaffold foo --path bar`
2. `yarn rw g scaffold bar/foo`

Currently, the command allows either b/c easy with Yargs. Personally I prefer number 2 from a "that just makes intuitive sense to write" perspective. **But I don't think that it's going to be that intuitive in context.** E.g. here's the current output for the `yarn rw generate --help`:
```bash
Commands:
  rw generate auth <provider>            Generates auth configuration.
  ...
  rw generate layout <name>              Generate a layout component.
  rw generate page <name> [path]         Generates a page component.
  rw generate scaffold <pathSlashModel>  Generate pages, SDL, and a services
                                         object.
  rw generate sdl <model>                Generate a GraphQL schema and service
                                         object.
  rw generate service <name>             Generate a service component.
```

- we don't actually get all the help text needed to explain what's going on
- it just sticks out like a sore thumb as "not being the same as the other"... so begs the 🤔
- it makes 'path' seem to be required

**==> Based on this, I'd recommend:**
- _removing_ the `<pathSlashModel>` option
- _using_ `rw generate scaffold <model> [path]`
- _change_ --help text for scaffold command to be `Generate Pages, SDL, and Services files based on a given DB schema Model`
- _add_ --help text for scaffold [path] option to be `nests files by type at the given path directory (or directories) `

**Comment by cannikin** [comment]
You know I'm always going to default the way Rails does it, and Rails uses the "just type out the path" version:

    yarn rw g scaffold admin/user

To me that's much clearer than:

    yarn rw g scaffold user admin

Which is "backwards" from the structure that's actually generated!

I'd just leave that option as `<model>` and then add the `--help` text that explains that you can create your scaffold in a subdirectory by just writing out the path you want to generate.

**Comment by thedavidprice** [comment]
>Which is "backwards" from the structure that's actually generated!

^^ 100% agreed, which is why I also preferred `yarn rw g scaffold admin/user`

>I'd just leave that option as <model>

^^ Ah, great idea. This would be ideal if Yargs will let us do it. 

So, to recap (and please edit as needed):
- the **command display** should be `rw generate scaffold <model>`
- the **scaffold help text** should be `Generate Pages, SDL, and Services files based on a given DB schema Model`
- the scaffold **option help text** should be `optionally may use "path/model" to nest files by type at the given path directory (or directories)`

@jtoar do you think this is possible?

**Comment by jtoar** [comment]
Totally; just put up #559. Just want to make sure I got the help message right:

<details><summary>Output for <code>yarn rw g --help</code></summary>

```
~/redwood-app$ yarn rw g --help
yarn run v1.22.4
$ /redwood-app/node_modules/.bin/rw g --help
rw generate <type>

Save time by generating boilerplate code.

Commands:
  rw g auth <provider>     Generates auth configuration.
  rw g cell <name>         Generate a cell component.
  rw g component <name>    Generate a component component.
  rw g function <name>     Generate a function
  rw g layout <name>       Generate a layout component.
  rw g page <name> [path]  Generates a page component.
  rw g scaffold <model>    Generate Pages, SDL, and Services files based on a
                           given DB schema Model.
  rw g sdl <model>         Generate a GraphQL schema and service object.
  rw g service <name>      Generate a service component.

Options:
  --help     Show help                                                 [boolean]
  --version  Show version number                                       [boolean]
```
</details>

<details><summary>Output for <code>yarn rw g scaffold --help</code></summary>

```
~/redwood-app$ yarn rw g scaffold --help
yarn run v1.22.4
$ /redwood-app/node_modules/.bin/rw g scaffold --help
rw g scaffold <model>

Generate Pages, SDL, and Services files based on a given DB schema Model.

Positionals:
  model  Model to scaffold. Optionally may use "path/model" to nest files by
         type at the given path directory (or directories)            [required]

Options:
  --help     Show help                                                 [boolean]
  --version  Show version number                                       [boolean]
  --force                                             [boolean] [default: false]
```
</details>

**Comment by antonmoiseev** [comment]
@jtoar  Sorry for raising this question a bit tool late after the work was already done. I was going to offer my help, but you were too quick! Thank you!




> Is the "it" you're referring to the Destroy command? If so, how would supporting --path differ from your current implementation.

@thedavidprice I was referring specifically to `destroy scaffold` command, but since we are dropping `--path`, I have no more questions. Thank you.

**Comment by jtoar** [comment]
@antonmoiseev No worries! It was a super simple fix so thank you for bringing it up--we can get it right before getting it out.

**Comment by thedavidprice** [comment]
@antonmoiseev Is there further work in progress here? If not, we are looking to release v0.7 tomorrow and it would be great to include this. But all good if need more time.

Thanks!

**Comment by antonmoiseev** [comment]
@thedavidprice Yep, I saw your RC release, so full steam ahead, aiming to push final tests in a couple of hours 😄. Hopefully it will make it into v0.7, but no worries if it won't be good enough or fully reviewed by tomorrow.

**Comment by antonmoiseev** [comment]
@cannikin It's ready for review, however as soon as #559 is merged, I'll need to adjust `destroy scaffold` command since the parameter names will change.

/cc @thedavidprice

**Comment by thedavidprice** [comment]
#559 has been merged

**Comment by antonmoiseev** [comment]
Adjusted to scaffold generator changes, merged with master, ready for review now.

**Comment by peterp** [comment]
Thanks for this! It looks good to me, but I'm going to check it out and test locally.

**Comment by peterp** [comment]
@antonmoiseev Just a heads up that I'm trying this locally and getting the following:
```terminal
yarn rw d cell Test
yarn run v1.22.4
$ /Users/peterp/x/redwoodjs/example-blog/node_modules/.bin/rw d cell Test
rw <command>

Commands:
  rw d cell <name>         Destroy a cell component.
  rw d component <name>    Destroy a component.
  rw d function <name>     Destroy a function
  rw d layout <name>       Destroy a layout component.
  rw d page <name> [path]  Destroy a page component.
  rw d scaffold <model>    Destroy pages, SDL, and a services object.
  rw d sdl <model>         Destroy a GraphQL schema and service object.
  rw d service <name>      Destroy a service component.

Options:
  --help     Show help                                                 [boolean]
  --version  Show version number                                       [boolean]

Unknown argument: Test```

I'm attempting to debug why this is hapenning.

File: packages/cli/src/commands/destroy.js [hunk 4]
```diff
@@ -0,0 +1,6 @@
+export const command = 'destroy <type>'
+export const aliases = ['d']
+export const desc = 'Rollback changes made by generate command.'
+
+export const builder = (yargs) =>
+  yargs.commandDir('./destroy', { recurse: true }).demandCommand().argv
```
Comment (peterp) [inline packages/cli/src/commands/destroy.js:6]: We changed some things in master, `argv` is no longer required.
```suggestion
  yargs.commandDir('./destroy', { recurse: true }).demandCommand()
```

File: packages/cli/src/commands/destroy.js [hunk 5]
```diff
@@ -0,0 +1,6 @@
+export const command = 'destroy <type>'
+export const aliases = ['d']
+export const desc = 'Rollback changes made by generate command.'
```
Comment (peterp) [inline packages/cli/src/commands/destroy.js:3]:
```suggestion
export const desc = 'Rollback changes made by the generate command.'
```

**Comment by antonmoiseev** [comment]
@peterp Thank you!

**Pull request merged**